docs(MADR): SPIFFE ID matches for MeshTrafficPermission (#13848)
kumahq/kuma: docs/madr/decisions/080-inbound-policies-matches.md

2. If the incoming request matches at least one matcher in either `allow` or `allowWithShadowDeny` list – the result is ALLOW.
3. If the incoming request doesn't match anything – the result is DENY.

#### Algorithm

1. Collect all MeshTrafficPermissions that target the inbound
2. Concat all `rules` arrays (we don't want to merge `default` confs to solve the [Problem 2](#problem-2))

```yaml
rules:
   - default: # from 'by-mesh-operator' MTP
       deny:
          - spiffeId:
              type: Exact
              value: "spiffe://trust-domain.mesh/ns/default/sa/frontend"
   - default: # from 'by-service-owner' MTP
        deny:
           - spiffeId:
                type: Exact
                value: "spiffe://trust-domain.mesh/ns/default/sa/api-gateway"
        allowWithShadowDeny:
           - spiffeId:
                type: Prefix
                value: "spiffe://trust-domain.mesh/ns/legacy"
        allow:
           - spiffeId:
                type: Prefix
                value: "spiffe://trust-domain.mesh/"
```

3. Starting this step MTP plugin is in charge, it receives the struct from step 2 and generates Envoy config.

> For each item in `rules` generate appropriate `matcher_list` and then concat `matcher_list` from old rules prioritizing `Deny`:

Author: ```suggestion
For each item in `rules` generate appropriate `matcher_list` and then concat `matcher_list` from all rules prioritizing `Deny`:
```

```yaml

> extensions.filters.network.rbac.v3.RBAC:
>   rules: []
>   shadow_rules: []
>   matcher:
>     matcher_list:
>        - predicate:
>            or_matcher:
>               - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/api-gateway
>            on_match:
>              action: Deny
>              name: kri_mtp_default___by-mesh-operator_
>        - predicate:
>             or_matcher:
>                - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/api-gateway
>             on_match:
>                action: Deny
>                name: kri_mtp_default___by-service-owner_
>        - predicate:
>             or_matcher:
>                - spiffeId prefix spiffe://trust-domain.mesh/
>                - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy # for 'allowWithShadowDeny'
>             on_match: Allow
>             name: kri_mtp_default___by-service-owner_
>     no_match: Deny
>   shadow_matcher: # not enforced, just logged
>     matcher_list:
>      - predicate:
>           or_matcher:
>              - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy
>           on_match: Deny
>           name: kri_mtp_default___by-service-owner_
>     no_match: Deny

Reviewer: Nit:
```suggestion
extensions.filters.network.rbac.v3.RBAC:
  rules: []
  shadow_rules: []
  matcher:
    matcher_list:
      - predicate:
          or_matcher:
            - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/api-gateway
          on_match:
            action: Deny
            name: kri_mtp_default___by-mesh-operator_
      - predicate:
          or_matcher:
            - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/api-gateway
          on_match:
            action: Deny
            name: kri_mtp_default___by-service-owner_
      - predicate:
          or_matcher:
            - spiffeId prefix spiffe://trust-domain.mesh/
            - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy # for 'allowWithShadowDeny'
          on_match: Allow
          name: kri_mtp_default___by-service-owner_
    no_match: Deny
  shadow_matcher: # not enforced, just logged
    matcher_list:
      - predicate:
          or_matcher:
            - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy
          on_match: Deny
          name: kri_mtp_default___by-service-owner_
    no_match: Deny
```

```

#### Inspect API

Inspect API for inbound rules:

```
GET :5681/meshes/{mesh}/dataplanes/{name}/_inbounds/{inbound-kri}/_policies
```
```yaml

> policies:
>   - kind: MeshTrafficPermission
>     rules:
>       - conf:
>           deny:
>              - spiffeId:
>                  type: Exact
>                  value: "spiffe://trust-domain.mesh/ns/default/sa/frontend"
>         origin: kri_mtp_default___by-mesh-operator_ # current Inspect API missing 'origin' inside each 'rule'
>       - conf:
>           deny:
>              - spiffeId:
>                   type: Exact
>                   value: "spiffe://trust-domain.mesh/ns/default/sa/api-gateway"
>           allowWithShadowDeny:
>              - spiffeId:
>                   type: Prefix
>                   value: "spiffe://trust-domain.mesh/ns/legacy"
>           allow:
>              - spiffeId:
>                   type: Prefix
>                   value: "spiffe://trust-domain.mesh/"
>         origin: kri_mtp_default___by-service-owner_
>     origins:
>        - kri: kri_mtp_...
>        - kri: kri_mtp_...

Reviewer: Nit:
```suggestion
policies:
  - kind: MeshTrafficPermission
    rules:
      - conf:
          deny:
            - spiffeId:
                type: Exact
                value: "spiffe://trust-domain.mesh/ns/default/sa/frontend"
        origin: kri_mtp_default___by-mesh-operator_ # current Inspect API missing 'origin' inside each 'rule'
      - conf:
          deny:
            - spiffeId:
                type: Exact
                value: "spiffe://trust-domain.mesh/ns/default/sa/api-gateway"
          allowWithShadowDeny:
            - spiffeId:
                type: Prefix
                value: "spiffe://trust-domain.mesh/ns/legacy"
          allow:
            - spiffeId:
                type: Prefix
                value: "spiffe://trust-domain.mesh/"
        origin: kri_mtp_default___by-service-owner_
    origins:
      - kri: kri_mtp_...
      - kri: kri_mtp_...
```

```

Currently, Inspect API is missing `policies[].rules[].origin`.

> As we don't want to merge `rules` to fix the [Problem 2](#problem-2) we can unambiguously match what policy KRI contributed what rule.

Author: https://github.com/kumahq/kuma/issues/14010

* Although the top-level targetRef supports `sectionName` and allows configuring a specific inbound,
it's not reflected in the Inspect API, since it reuses the proxy policy endpoint, which doesn't expose per-inbound configuration

#### Schema

The idea is to make MeshTrafficPermission a simple single-item policy:

```yaml
type: MeshTrafficPermission
mesh: default
name: by-mesh-operator
spec:
   targetRef: {}
   default:
      deny:
         - spiffeId:
              type: Exact
              value: "spiffe://trust-domain.mesh/ns/default/sa/frontend"
---
type: MeshTrafficPermission
mesh: default
name: by-service-owner
spec:
  targetRef:
     kind: Dataplane
     labels:
       app: backend
     sectionName: http-port
  default:
    deny:
      - spiffeId:
           type: Exact
           value: "spiffe://trust-domain.mesh/ns/default/sa/api-gateway"
    allowWithShadowDeny:
       - spiffeId:
           type: Prefix
           value: "spiffe://trust-domain.mesh/ns/legacy"
    allow:
       - spiffeId:
            type: Prefix
            value: "spiffe://trust-domain.mesh/"
```

Evaluation rules:

> 1. If the incoming request matches at least one matcher in `deny` list – the result is DENY.
> 2. If the incoming request matches at least one matcher in either `allow` or `allowWithShadowDeny` list – the result is ALLOW.
> 3. If the incoming request doesn't match anything – the result is DENY.

Reviewer: Nit:
```suggestion
1. If the incoming request matches at least one matcher in `deny` list - the result is DENY.
2. If the incoming request matches at least one matcher in either `allow` or `allowWithShadowDeny` list - the result is ALLOW.
3. If the incoming request doesn't match anything - the result is DENY.
```

Author: I'd prefer en-dash instead of a hyphen

#### Merging

Normally in Kuma policies mergeable arrays override each other, for example:

> ```

Reviewer: Nit:
````suggestion
```yaml
````

name: policy-1
spec:
  default:
    myArray: [1,2,3]
---
name: policy-1
spec:
  default:
    myArray: [4]
```

results in `spec.default.myArray: [4]`.

Switching array merging behaviour from overriding to concatenation normally requires prefixing fields with `append`:

> ```

Reviewer: Nit:
````suggestion
```yaml
````

name: policy-1
spec:
  default:
    appendArray: [1,2,3]
---
name: policy-1
spec:
  default:
    appendArray: [4]
```

results in `spec.default.appendArray: [1,2,3,4]`.

Instead of naming fields `appendDeny` and `appendAllow` we need to add a new value `concat` for the existing `policyMerge` struct tag:

```go
type Conf struct {
    Allow *[]Matcher `json:"allow,omitempty" policyMerge:"concat"`
}
```

When used on fields prefixed with `append` the struct tag `concat` won't have any additional effect.

Kubernetes has similar approach in their API with [merge-strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy).
We don't have to adopt their exact annotations as we won't gain anything from it.
But it shows that the concept is viable,
and potentially we can update `MeshPassthrough`, `MeshProxyPatch` and `MeshOPA` to use struct tags instead of `append` prefix.

#### Algorithm

1. Collect all MeshTrafficPermissions that target the inbound
2. Concat all `deny`, `allowWithShadowDeny` and `allow` lists
3. Convert to the following Matching API structure

```yaml

> extensions.filters.network.rbac.v3.RBAC:
>   rules: []
>   shadow_rules: []
>   matcher:
>     matcher_list:
>        - predicate:
>            or_matcher:
>               - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/frontend
>               - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/api-gateway
>            on_match:
>              action: Deny
>        - predicate:
>             or_matcher:
>                - spiffeId prefix spiffe://trust-domain.mesh/
>                - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy # for 'allowWithShadowDeny'
>             on_match: Allow
>     no_match: Deny
>   shadow_matcher: # not enforced, just logged
>     matcher_list:
>      - predicate:
>           or_matcher:
>              - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy
>           on_match: Deny
>     no_match: Deny

Reviewer: Nit:
```suggestion
extensions.filters.network.rbac.v3.RBAC:
  rules: []
  shadow_rules: []
  matcher:
    matcher_list:
      - predicate:
          or_matcher:
            - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/frontend
            - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/api-gateway
          on_match:
            action: Deny
      - predicate:
          or_matcher:
            - spiffeId prefix spiffe://trust-domain.mesh/
            - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy # for 'allowWithShadowDeny'
          on_match: Allow
    no_match: Deny
  shadow_matcher: # not enforced, just logged
    matcher_list:
      - predicate:
          or_matcher:
            - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy
          on_match: Deny
    no_match: Deny
```

```

#### Inspect API

Same as single-item policies (we call them proxy policy in the new Inspect API).

```
GET :5681/meshes/{mesh}/dataplanes/{name}/_policies
```
```yaml

> policies:
>    - kind: MeshTrafficPermission
>      conf:
>        deny: [...]
>        allowWithShadowDeny: [...]
>        allow: [...]
>      origins:
>         - kri: kri_mtp_...
>         - kri: kri_mtp_...

Reviewer: Nit:
```suggestion
policies:
  - kind: MeshTrafficPermission
    conf:
      deny: [...]
      allowWithShadowDeny: [...]
      allow: [...]
    origins:
      - kri: kri_mtp_...
      - kri: kri_mtp_...
```

```

#### Verify user stories

##### Mesh Operator

1. I want all requests in the mesh to be denied by default (2.12)

> No MeshTrafficPermission policies means requests are denied by default.

Reviewer: Nit:
```suggestion
   No MeshTrafficPermission policies means requests are denied by default.
```

2. I want to declare a group of identities as explicitly denied,
   so that Service Owner can't override or bypass that decision,
   ensuring enforcement of critical security boundaries across the mesh. (2.12)

> ```yaml
> type: MeshTrafficPermission
> mesh: default
> name: by-mesh-operator
> spec:
>    default:
>       deny:
>          - spiffeId:
>               type: Exact
>               value: "spiffe://trust-domain.mesh/ns/default/sa/api-gateway"
>          - spiffeId:
>               type: Exact
>               value: "spiffe://trust-domain.mesh/ns/default/sa/legacy-workload"
>          - spiffeId:
>               type: Prefix
>               value: "spiffe://legacy.mesh/"
> ```

Reviewer: Nit:
````suggestion
   ```yaml
   type: MeshTrafficPermission
   mesh: default
   name: by-mesh-operator
   spec:
     default:
       deny:
         - spiffeId:
             type: Exact
             value: "spiffe://trust-domain.mesh/ns/default/sa/api-gateway"
         - spiffeId:
             type: Exact
             value: "spiffe://trust-domain.mesh/ns/default/sa/legacy-workload"
         - spiffeId:
             type: Prefix
             value: "spiffe://legacy.mesh/"
   ```
````

3. I want to allow all clients in the `observability` namespace to access all services by default,
   so that telemetry and monitoring tools function automatically,
   while still allowing Service Owners to explicitly opt out by applying deny policies. (2.12)

> ```yaml
> type: MeshTrafficPermission
> mesh: default
> name: by-mesh-operator
> spec:
>    default:
>       allow:
>          - spiffeId:
>               type: Prefix
>               value: "spiffe://trust-domain.mesh/ns/observability"
> ```

Reviewer: Nit:
````suggestion
   ```yaml
   type: MeshTrafficPermission
   mesh: default
   name: by-mesh-operator
   spec:
     default:
       allow:
         - spiffeId:
             type: Prefix
             value: "spiffe://trust-domain.mesh/ns/observability"
     ```
````

4. I want to allow all clients in the `observability` namespace to access the `/metrics` endpoint on all services,
   so that monitoring tools can collect metrics without requiring each Service Owner to configure access individually.

> ```yaml
> type: MeshTrafficPermission
> mesh: default
> name: by-mesh-operator
> spec:
>    default:
>       allow:
>          - spiffeId:
>               type: Prefix
>               value: "spiffe://trust-domain.mesh/ns/observability"
>            path:
>              type: Prefix
>              value: "/metrics"
> ```

Reviewer: Nit:
````suggestion
   ```yaml
   type: MeshTrafficPermission
   mesh: default
   name: by-mesh-operator
   spec:
     default:
       allow:
         - spiffeId:
             type: Prefix
             value: "spiffe://trust-domain.mesh/ns/observability"
           path:
             type: Prefix
             value: "/metrics"
     ```
````

##### Service Owner

1. I want to grant access to my service to any client I choose,
   so that I can support integrations and collaboration with other teams,
   unless the Mesh Operator has explicitly denied that client,
   ensuring I remain in control while respecting mesh-wide security boundaries. (2.12)

> ```yaml
> type: MeshTrafficPermission
> mesh: default
> name: by-backend-owner
> spec:
>    targetRef:
>       kind: Dataplane
>       labels:
>          app: backend
>    default:
>       allow:
>          - spiffeId:
>               type: Prefix
>               value: "spiffe://trust-domain.mesh/"
> ```

Reviewer: Nit:
````suggestion
   ```yaml
   type: MeshTrafficPermission
   mesh: default
   name: by-backend-owner
   spec:
     targetRef:
       kind: Dataplane
       labels:
         app: backend
     default:
       allow:
         - spiffeId:
             type: Prefix
             value: "spiffe://trust-domain.mesh/"
     ```
````

                type: Prefix
                value: "spiffe://trust-domain.mesh/"
             method: GET
        default: $conf
```

MeshTrafficPermission uses `conditions` instead of `matches` to emphasize ALL of them need to be evaluated:

```yaml
type: MeshTrafficPermission
mesh: default
spec:
   rules:
      - conditions:
           - spiffeId:
                type: Exact
                value: "spiffe://trust-domain.mesh/ns/default/sa/frontend"
        default:
           action: Deny
---
type: MeshTrafficPermission
mesh: default
spec:
  rules:
    - conditions:
         - spiffeId:
              type: Exact
              value: "spiffe://trust-domain.mesh/ns/default/sa/api-gateway"
      default:
        action: Deny
    - conditions:
         - spiffeId:
              type: Prefix
              value: "spiffe://trust-domain.mesh/"
      default:
        action: Allow
    - conditions:
         - spiffeId:
              type: Prefix
              value: "spiffe://trust-domain.mesh/ns/legacy"
      default:
         action: AllowWithShadowDeny
```

#### Algorithm

1. Collect all MeshTrafficPermissions that target the inbound
2. Concat all `rules` lists

> 3. Without peeking into confs, group `conditions` by the same confs (either compare conf's hashes or marshalled content).
> With MTP we're getting at most 3 unique confs – `action: Deny`, `action: AllowWithShadowDeny` and `action: Allow`

Reviewer: Nit: Wdyt?
```suggestion
3. Group conditions by identical confs without inspecting them directly (use either their hash or marshalled content). With MeshTrafficPermission, there can be at most 3 unique confs: `action: Deny`, `action: AllowWithShadowDeny`, and `action: Allow`.
```

   unless the Mesh Operator has explicitly denied that client,
   ensuring I remain in control while respecting mesh-wide security boundaries. (2.12)

2. I want to opt out of mesh-wide `observability` access,
   by denying requests from the `observability` namespace,
   so that my service’s sensitive endpoints remain private unless explicitly allowed. (2.12)

3. I want to block malicious/abusive client even if previously it was allowed,
   so I can prevent my service from overloading until client's service team reacts to the incident. (2.12)

4. I want to allow `GET` requests to my service from any client, but restrict `POST` requests to a group of identities,
   so that read operations are public but write operations are gated.

### Design

According to [MADR-078](078-special-mtp-algo.md) MeshTrafficPermission's algorithm and API is different from other inbound policies.
Current MADR suggests 2 options to implement the design. 
Chosen option is "Option 1: MeshTrafficPermission is a single-item policy".

#### Option 1: MeshTrafficPermission is a single-item policy

Pros:
* Inspect API works out-of-the-box!
* No need to introduce a new type of policy

> * `allowRules`, `denyRules` reuse `rules[].matches[]` schema

Reviewer: Maybe it would be good to link to this schema. It might be a link to the documentation or the source code. Wdyt?

Author: Unfortunately there is no schema for inbound policy matches yet

> * No intermediate representation is required, `rbac_configurer.go` generates Envoy configuration directly from `conf`

Reviewer: You should explain what does intermediate representation means and be more specific when mentioning `conf` (what is conf, where it comes from etc.)

   so that monitoring tools can collect metrics without requiring each Service Owner to configure access individually.

```yaml
type: MeshTrafficPermission
mesh: default
name: by-mesh-operator
spec:
   default:
      allow:
         - spiffeId:
              type: Prefix
              value: "spiffe://trust-domain.mesh/ns/observability"
           path:
             type: Prefix
             value: "/metrics"
```

###### Service Owner

1. I want to grant access to my service to any client I choose,
   so that I can support integrations and collaboration with other teams,
   unless the Mesh Operator has explicitly denied that client,
   ensuring I remain in control while respecting mesh-wide security boundaries. (2.12)

```yaml
type: MeshTrafficPermission
mesh: default
name: by-backend-owner
spec:
   targetRef:
      kind: Dataplane
      labels:
         app: backend
   default:
      allow:
         - spiffeId:
              type: Prefix
              value: "spiffe://trust-domain.mesh/"
```

2. I want to opt out of mesh-wide `observability` access,
   by denying requests from the `observability` namespace,

>    so that my service’s sensitive endpoints remain private unless explicitly allowed. (2.12)

Reviewer 2: Does this mean `spiffe://trust-domain.mesh/ns/observability` can still access `/metrics` or not? 🤔

   so that telemetry and monitoring tools function automatically,
   while still allowing Service Owners to explicitly opt out by applying deny policies. (2.12)

4. I want to be able to allow all clients in the `observability` namespace to access the `/metrics` endpoint on all services,
   so that monitoring tools can collect metrics without requiring each Service Owner to configure access individually.

#### Service Owner

1. I want to be able to grant access to my service to any client I choose,
   so that I can support integrations and collaboration with other teams,
   unless the Mesh Operator has explicitly denied that client,
   ensuring I remain in control while respecting mesh-wide security boundaries. (2.12)

2. I want to be able to opt out of mesh-wide `observability` access,
   by denying requests from the `observability` namespace,
   so that my service’s sensitive endpoints remain private unless explicitly allowed. (2.12)

3. I want to be able to block malicious/abusive client even if previously it was allowed,
   so I can prevent my service from overloading until the client's service team responds to the incident (2.12)

> 4. I want to be able to allow `GET` requests to my service from any client, but restrict `POST` requests to a group of identities,

Author: You're right, added new user story and added new option (it's option 1 now) that covers the new user story as well

##### Merging

Normally in Kuma policies mergeable arrays override each other, for example:

```
name: policy-1
spec:
  default:
    myArray: [1,2,3]
---
name: policy-1
spec:
  default:
    myArray: [4]
```

results in `spec.default.myArray: [4]`.

Switching array merging behaviour from overriding to concatenation normally requires prefixing fields with `append`:

```
name: policy-1
spec:
  default:
    appendArray: [1,2,3]
---
name: policy-1
spec:
  default:
    appendArray: [4]
```

results in `spec.default.appendArray: [1,2,3,4]`.

Instead of naming fields `appendDeny` and `appendAllow` we need to add a new value `concat` for the existing `policyMerge` struct tag:

```go
type Conf struct {
    Allow *[]Matcher `json:"allow,omitempty" policyMerge:"concat"`
}
```

When used on fields prefixed with `append` the struct tag `concat` won't have any additional effect.

Kubernetes has similar approach in their API with [merge-strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy).
We don't have to adopt their exact annotations as we won't gain anything from it.
But it shows that the concept is viable

> and potentially we can update `MeshPassthrough`, `MeshProxyPatch` and `MeshOPA` to use struct tags instead of `append` prefix.

Reviewer 2: I think that we would keep the old fields (`appendDeny`, `appendAllow`) and have `policyMerge` on new fields (`deny`, `allow`).

> and potentially we can update `MeshPassthrough`, `MeshProxyPatch` and `MeshOPA` to use struct tags instead of `append` prefix.

Reviewer 3: yes from now on we could do this and keep the old stuff as is

  rules: []
  shadow_rules: []
  matcher:
    matcher_list:
       - predicate:
           or_matcher:
              - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/frontend
              - spiffeId exact spiffe://trust-domain.mesh/ns/default/sa/api-gateway
           on_match:
             action: Deny
             name: kri_mtp_
       - predicate:
            or_matcher:
               - spiffeId prefix spiffe://trust-domain.mesh/
               - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy # for 'allowWithShadowDeny'
            on_match: Allow
    no_match: Deny
  shadow_matcher: # not enforced, just logged
    matcher_list:
     - predicate:
          or_matcher:
             - spiffeId prefix spiffe://trust-domain.mesh/ns/legacy
          on_match: Deny
    no_match: Deny
```

##### Inspect API

Same as singe-item policies (we call them proxy policy in the new Inspect API).

```
GET :5681/meshes/{mesh}/dataplanes/{name}/_policies
```
```yaml
policies:
   - kind: MeshTrafficPermission
     conf:
       deny: [...]
       allowWithShadowDeny: [...]
       allow: [...]
     origins:
        - kri: kri_mtp_...
        - kri: kri_mtp_...
```

> ##### Extensibility with DPP labels selector

Author: Updated the section

2. If the incoming request matches at least one matcher in either `allow` or `allowWithShadowDeny` list – the result is ALLOW.
3. If the incoming request doesn't match anything – the result is DENY.

#### Algorithm

1. Collect all MeshTrafficPermissions that target the inbound
2. Concat all `rules` arrays (we don't want to merge `default` confs to solve the [Problem 2](#problem-2))

```yaml
rules:
   - default: # from 'by-mesh-operator' MTP
       deny:
          - spiffeId:
              type: Exact
              value: "spiffe://trust-domain.mesh/ns/default/sa/frontend"
   - default: # from 'by-service-owner' MTP
        deny:
           - spiffeId:
                type: Exact
                value: "spiffe://trust-domain.mesh/ns/default/sa/api-gateway"
        allowWithShadowDeny:
           - spiffeId:
                type: Prefix
                value: "spiffe://trust-domain.mesh/ns/legacy"
        allow:
           - spiffeId:
                type: Prefix
                value: "spiffe://trust-domain.mesh/"
```

3. Starting this step MTP plugin is in charge, it receives the struct from step 2 and generates Envoy config.

> For each item in `rules` generate appropriate `matcher_list` and then concat `matcher_list` from old rules prioritizing `Deny`:

Reviewer 2: > old rules prioritizing `Deny`

is this easy to do?

Author: I don't know what happened here, but I wanted to type "all rules" not "old rules" 🤦‍♂️

            value: "spiffe://trust-domain.mesh/"
```

##### Merging

Normally in Kuma policies mergeable arrays override each other, for example:

```
name: policy-1
spec:
  default:
    myArray: [1,2,3]
---
name: policy-1
spec:
  default:
    myArray: [4]
```

results in `spec.default.myArray: [4]`.

Switching array merging behaviour from overriding to concatenation normally requires prefixing fields with `append`:

```
name: policy-1
spec:
  default:
    appendArray: [1,2,3]
---
name: policy-1
spec:
  default:
    appendArray: [4]
```

results in `spec.default.appendArray: [1,2,3,4]`.

> Instead of naming fields `appendDeny` and `appendAllow` we need to add a new value `concat` for the existing `policyMerge` struct tag:

Reviewer 2: Is aesthetics the only reason for this? Honestly I don't mind it being called `append*` - that way the api is consistent (no need to explain to the user, there are some fields that append that have `append` prefix and some don't)

Author: Discussed offline to go with struct tag instead of `append`

```yaml
type: MeshTrafficPermission
mesh: default
name: by-backend-owner
spec:
   targetRef:
      kind: Dataplane
      labels:
         app: backend
   default:
      deny:
         - spiffeId:
              type: Prefix
              value: "spiffe://trust-domain.mesh/ns/observability"
```

3. I want to block malicious/abusive client even if previously it was allowed,
   so I can prevent my service from overloading until client's service team reacts to the incident. (2.12)

```yaml
type: MeshTrafficPermission
mesh: default
name: by-backend-owner
spec:
   targetRef:
      kind: Dataplane
      labels:
         app: backend
   default:
      deny:
         - spiffeId:
              type: Exact
              value: "spiffe://trust-domain.mesh/ns/default/sa/malicious"
```

4. I want to allow `GET` requests to my service from any client, but restrict `POST` requests to a group of identities,
   so that read operations are public but write operations are gated.

```yaml
type: MeshTrafficPermission
mesh: default
name: by-backend-owner
spec:
   targetRef:
      kind: Dataplane
      labels:
         app: backend
   default:
      allow:
         - method: GET

>       deny:

Reviewer 2: is this right? If everything is deny by default you just added a more detailed deny list 🤔

Author: It's not right, I should've added these POST matches to `allow`, let me fix the example

        - kri: kri_mtp_...
```

> ##### Extensibility with DPP labels selector

Reviewer 2: should we have mention why we're removing this?

Reviewer 2: explained in https://github.com/kumahq/kuma/pull/13848/files#diff-45dce14e9572bcf179427f441dbe76eba3742cf03befd54964e96b41e37847e0R468
